### ISSUE TITLE
MObjectHandle error when trying to select a custom plugin node

### ISSUE THREAD

**Issue by elrond79** [LumaPictures/pymel#186]
_From [drewskil...@gmail.com](https://code.google.com/u/102793369357456976667/) on January 19, 2010 17:37:13_

When I try and select a custom plugin node, I get this:

for x in ls( type='transform'):
    print x.longName()

\# Error: 'MObjectHandle'
\# Traceback (most recent call last):
\#   File "<maya console>", line 2, in \<module>
\#   File "c:\dfp\Common\Code\Tools\Maya\Python\pymel-
1.0.0rc1\pymel\core\nodetypes.py", line 297, in __getattr__
\#     return DependNode.attr(self,attr)
\#   File "c:\dfp\Common\Code\Tools\Maya\Python\pymel-
1.0.0rc1\pymel\core\nodetypes.py", line 382, in attr
\#     plug = self.__apimfn__().findPlug( attr, False )
\#   File "c:\dfp\Common\Code\Tools\Maya\Python\pymel-
1.0.0rc1\pymel\core\general.py", line 1661, in __apimfn__
\#     obj = self.__apiobject__()
\#   File "c:\dfp-stack\Common\Code\Tools\Maya\Python\pymel-
1.0.0rc1\pymel\core\nodetypes.py", line 119, in __apiobject__
\#     return self.__apimobject__()
\#   File "c:\dfp\Common\Code\Tools\Maya\Python\pymel-
1.0.0rc1\pymel\core\nodetypes.py", line 123, in __apimobject__
\#     handle = self.__apihandle__()
\#   File "c:\dfp\Common\Code\Tools\Maya\Python\pymel-
1.0.0rc1\pymel\core\nodetypes.py", line 129, in __apihandle__
\#     return self.__apiobjects__['MObjectHandle']
\# KeyError: 'MObjectHandle' # What Version of Maya are you using? On what operating system? (Please be as specific as possible and include service packs, 64bit vs 32bit, etc) 2008 windows 32bit, latest service pack What Version of PyMEL are you using? 1.0.0rc1 What is the expected output? What do you see instead? If possible, provide a few lines of code to reproduce the problem from scratch. As seen above, expected to list a pymel node for my object Is there a maya.cmds or maya.OpenMaya equivalent to the pymel function in question? ( For example, pymel.listRelatives vs. maya.cmds.listRelatives ) If so, does it exhibit the same problem? No, If I import maya.cmds as * instead of pymel.cmds as *, this works as 
expected:
for x in ls( type='transform'):
    print x
\>>Room_SetDressing
\>>Sphere_SetDressing
\>>... Please provide any additional information below. I'm afraid I can't provide many details on the custom plugin we're using, 
as I'm not familiar with this system.  But I can get a member of our tools 
department to describe additional details.  Also note the original thread 
on the issue here: http://tech-artists.org/forum/showthread.php?t=675

_Original issue: http://code.google.com/p/pymel/issues/detail?id=151_

**Comment by elrond79** [comment]
_From [chad...@gmail.com](https://code.google.com/u/101505156600271307269/) on January 19, 2010 20:30:39_

we have confirmed that pymel works with custom nodes, so it would really help us out 
if they could put together a simple plugin that reproduces the problem.  otherwise, we 
don't have much to go on.

**Status:** Accepted  
  **Labels:** Priority-Critical Milestone-1.0.x

**Comment by elrond79** [comment]
_From [drewskil...@gmail.com](https://code.google.com/u/102793369357456976667/) on January 28, 2010 14:08:47_

Our lead tools programmer dug into the problem and found a fix for us. Here's my 
somewhat limited understanding of the problem:

We have a custom transform type "dfTerrainRelativeTransform" defined through our 
level editing plugin.  PyMel doesn't seem to currently support custom transform node 
types, but we were able to get things working by remapping our node type to the usual 
one in general.py:


++_dfNodeRemap = {
++        "dfTerrainRelativeTransform" : "Transform",
++}

def _getPymelTypeFromObject(obj, name):
        compTypes = _factories.apiEnumsToPyComponents.get(obj.apiType(), None)
        if compTypes is not None:
            if len(compTypes) == 1:
                return compTypes[0]
            else:
                raise RuntimeError('Got an instance of a component with more than one 
possible PyNode type: %s' % obj.apiTypeStr())
        else:
            try:  
                fnDepend = _api.MFnDependencyNode( obj )
                mayaType = fnDepend.typeName()
++              mayaType = _dfNodeRemap.get(mayaType,mayaType)
                import nodetypes
                pymelType = getattr( nodetypes, _util.capitalize(mayaType), 
nodetypes.DependNode )

**Comment by elrond79** [comment]
_From [chad...@gmail.com](https://code.google.com/u/101505156600271307269/) on March 10, 2010 09:06:25_

from Dave Dixon: 
"you should be able to reproduce the problem with any plugin that creates a new node
derived from MPxTransform. It seems like the problem occurs when you have a node
(that is not a shape node) in the selection list whose type does not map to one of
the pre-defined types you have in pymel. As an aside, it seems like the fallback case
of instantiating a DependNode wrapper class was causing some issues because it didn’t
ever get it’s maya object set (sorry, I can’t remember the exact name of the python
instance var since I don’t have the code in front of me)…"

**Comment by elrond79** [comment]
_From [chad...@gmail.com](https://code.google.com/u/101505156600271307269/) on April 10, 2010 21:06:08_

as i suspected, it's not quite that simple.  loading and unloading rockingTransform.py in the plugin devkit works 
fine.  are you loading your plugins in userSetup.py or userSetup.mel?

**Comment by elrond79** [comment]
_From [elron...@gmail.com](https://code.google.com/u/112796493857278763022/) on January 20, 2011 15:34:35_

looked into this again... and was able to load/unload/select/print PyNodes for rockingTransforms, stereoRigTransforms, and footPrintLocators... so either whatever bug there was, is now fixed, or there must be something more specific to the original poster's plug-in, and we can't proceed until we get more info.

Going to close this, unless we can get more info

**Status:** WontFix

**Comment by elrond79** [comment]
_From [drewskil...@gmail.com](https://code.google.com/u/102793369357456976667/) on January 20, 2011 15:48:03_

Sounds good, close away! Sorry I couldn't provide you guys with a version of the plugin and more details, but thanks for all the help. When we upgrade to the latest maya I'll go ahead and verify that the problem no longer occurs.

**Comment by elrond79** [comment]
_From [pierre.a...@gmail.com](https://code.google.com/u/116432443266570551407/) on April 15, 2011 02:06:32_

I confirm that this issue is still present in maya2011 and pymel 1.0.2
( And above, see ticket `#243` )

If a listHistory / listFuture is called on a node that have connections to custom nodes (PSD, shaveNode, ...) the same exception is raised.

Traceback (most recent call last):
  File "S:\python-2_6\delacave\Maya\lib\io\bake.py", line 187, in loadMeshCacheFile
    histo = pmnode.listFuture(type="historySwitch")
  File "S:\python-2_6\lib\pymel\core\general.py", line 801, in listFuture
    return listHistory(*args, **kwargs)
  File "S:\python-2_6\lib\pymel\core\general.py", line 784, in listHistory
    results = [x for x in results if type in x.nodeType(inherited=True)]
  File "S:\python-2_6\lib\pymel\internal\pmcmds.py", line 89, in wrappedCmd
    new_args = getMelRepresentation(args)
  File "S:\python-2_6\lib\pymel\internal\pmcmds.py", line 67, in getMelRepresentation
    newargs[index] = value.__melobject__()
  File "S:\python-2_6\lib\pymel\core\general.py", line 1718, in __melobject__
    return self.name()
  File "S:\python-2_6\lib\pymel\core\nodetypes.py", line 76, in name
    return self._updateName()
  File "S:\python-2_6\lib\pymel\core\nodetypes.py", line 65, in _updateName
    self.__apimobject__()
  File "S:\python-2_6\lib\pymel\core\nodetypes.py", line 123, in __apimobject__
    handle = self.__apihandle__()
  File "S:\python-2_6\lib\pymel\core\nodetypes.py", line 129, in __apihandle__
    return self.__apiobjects__['MObjectHandle']
KeyError: 'MObjectHandle'

**Comment by elrond79** [comment]
_From [aloys.ba...@gmail.com](https://code.google.com/u/100305154637344008745/) on September 05, 2011 18:15:56_

Hi,

We are experiencing the same exact issue with custom nodes. Of course this is an intermittent problem, and only happens in specific scenario involving custom nodes and complex setup that we can't share here.
It also looks like a load order problem. Loading custom plugins before PyMel sometimes fixes the problem, but not always.

18:41:14|   File "/usr/autodesk/maya2011.5-x64/lib/python2.6/site-packages/pymel/core/nodetypes.py", line 297, in __getattr__
18:41:14|     return DependNode.attr(self,attr)
18:41:14|   File "/usr/autodesk/maya2011.5-x64/lib/python2.6/site-packages/pymel/core/nodetypes.py", line 382, in attr
18:41:14|     plug = self.__apimfn__().findPlug( attr, False )
18:41:14|   File "/usr/autodesk/maya2011.5-x64/lib/python2.6/site-packages/pymel/core/general.py", line 1673, in __apimfn__
18:41:14|     obj = self.__apiobject__()
18:41:14|   File "/usr/autodesk/maya2011.5-x64/lib/python2.6/site-packages/pymel/core/nodetypes.py", line 119, in __apiobject__
18:41:14|     return self.__apimobject__()
18:41:14|   File "/usr/autodesk/maya2011.5-x64/lib/python2.6/site-packages/pymel/core/nodetypes.py", line 123, in __apimobject__
18:41:14|     handle = self.__apihandle__()
18:41:14|   File "/usr/autodesk/maya2011.5-x64/lib/python2.6/site-packages/pymel/core/nodetypes.py", line 129, in __apihandle__
18:41:14|     return self.__apiobjects__['MObjectHandle']
18:41:14| KeyError: 'MObjectHandle'

It looks like the __apihandle__ method of the DagNode class is protected against this particular issue, whereas the __apihandle__ method of DependNode, which is failing here, is not.
I will try to see if I can patch our pymel version to avoid the error.

We are on 2011.5 with the pymel version shipped by Autodesk, which looks like 1.0.0.

Cheers,

Aloys

**Comment by elrond79** [comment]
_From [aloys.ba...@gmail.com](https://code.google.com/u/100305154637344008745/) on September 05, 2011 18:47:16_

Interestingly, manually running the pymel.core._installCallbacks function before using pymel with custom nodes works.
I guess somehow Maya is not calling the callbacks in all circumstances.

Cheers,

Aloys
